image.c

```diff
@@ -133,6 +143,141 @@ openImgdisplay()
     return FALSE;
 }
 
+static int
+protocol_test_environment(void)
+{
+    const char * envptr;
+
+    if ((envptr = getenv("TERM"))) {
+	/* The purpose of strncmp(3) where used below is like the glob `prefix*',
+	 * or the regex `^prefix' */
+	if (strcmp(envptr, "xterm-kitty") == 0)
+	    return INLINE_IMG_KITTY;
+	if (strcmp(envptr, "xterm-ghostty") == 0)
+	    return INLINE_IMG_KITTY;
+	if (strncmp(envptr, "mlterm", 6) == 0)
+	    return INLINE_IMG_OSC5379;
+	/* yaft doesn't correctly respond to \e[c, but is sixel-capable
+	 * anyway. Thanks to hackerb9/lsix */
+	if (strncmp(envptr, "yaft", 4) == 0)
+	    return INLINE_IMG_SIXEL;
+    }
+
+    if ((envptr = getenv("KONSOLE_VERSION")) && strcmp(envptr, "220770") >= 0)
+	return INLINE_IMG_KITTY;
+
+    return INLINE_IMG_NONE;
+}
+
+static int
+have_img2sixel(void)
+{
+    pid_t child_pid;
+    int wstatus;
+
+    if (getenv("W3M_IMG2SIXEL"))
+	return TRUE;
+
+    switch (child_pid = fork()) {
+    case -1:
+	return FALSE;
+    case 0:
+	close(STDOUT_FILENO);
+	close(STDERR_FILENO);
+	execlp("img2sixel", "img2sixel", "--version", NULL);
+	/* if exec fails */
+	_exit(-1);
+    default:
+	return
+#ifdef HAVE_WAITPID
+	    waitpid(child_pid, &wstatus, 0) != -1
+#else
+	    wait(&wstatus) != -1
+#endif
+	    && WIFEXITED(wstatus)
+	    && WEXITSTATUS(wstatus) == 0;
+    }
+}
+
+/* NB: in theory, user input could be snarfed up with the read(2); in
+ * practice, only museum pieces have such low latency, so we should get
+ * the device attributes (and only the device attributes) immediately. If
+ * ever this becomes an issue (which it won't), ungetch(3) can be used on
+ * any extraneous input */
+static int
+protocol_test_for_sixel(void)
+{
+    static const char errstr[] = "Can't get terminal attributes";
+
+    /* the string in the following sizeof expression is what I think the
+     * largest possible response could be. See
+     *	https://invisible-island.net/xterm/ctlseqs/ctlseqs.html
+     * for more information, specifically the section on CSI Ps c ("\e[c") */
+    char term_response[sizeof "\033[?65;1;2;3;4;6;8;9;15;16;17;18;21;22;28;29c"] = "";
+    char * ptr; /* general-purpose pointer to the above buffer */
+    ssize_t nchars_read;
+
+    /* request tty send primary device attributes */
+    write(STDOUT_FILENO, "\033[c", 3);
+
+    /* wait for stdin to become available; don't let us get stuck hanging
+     * indefinitely on a response that won't come */
+    {
+	int nret;
+	fd_set fds;
+	struct timeval timeout = { 0, 0.2 * 1000000 }; /* wait 0.2s, from terms.c */
+	FD_ZERO(&fds);
+	FD_SET(STDIN_FILENO, &fds);
+	if ((nret = select(STDIN_FILENO + 1, &fds, NULL, NULL, &timeout)) <= 0) {
+	    fprintf(stderr, "%s: %s\n", errstr, nret == 0 ? "timed out" : strerror(errno));
```

Comment:
Please keep it below 80 chars, if there is no very good reason to do otherwise.

Applies to other spots as well.